addons/controllers/utils/utils.go

```diff
@@ -11,28 +11,80 @@ import (
+	labelMatch, err := labels.NewRequirement(clusterapiv1beta1.ClusterLabelName, selection.Equals, []string{cluster.Name})
+	if err != nil {
+		return nil, err
+	}
+	labelSelector := labels.NewSelector()
+	labelSelector = labelSelector.Add(*labelMatch)
+	if err := clt.List(ctx, vsphereClusters, &client.ListOptions{LabelSelector: labelSelector, Namespace: cluster.Namespace}); err != nil {
+		return nil, err
+	}
+	if len(vsphereClusters.Items) != 1 {
+		return nil, fmt.Errorf("expected to find 1 VSphereCluster object for label key %s and value %s but found %d",
+			clusterapiv1beta1.ClusterLabelName, cluster.Name, len(vsphereClusters.Items))
+	}
+	return &vsphereClusters.Items[0], nil
+}
+
+// GetVSphereClusterNonParavirtual gets the VSphereCluster CR for the cluster object in non-paravirtual mode
+func GetVSphereClusterNonParavirtual(ctx context.Context, clt client.Client, cluster *clusterapiv1beta1.Cluster) (*capvv1beta1.VSphereCluster, error) {
+
+	vsphereClusters := &capvv1beta1.VSphereClusterList{}
+	labelMatch, err := labels.NewRequirement(clusterapiv1beta1.ClusterLabelName, selection.Equals, []string{cluster.Name})
+	if err != nil {
+		return nil, err
+	}
+	labelSelector := labels.NewSelector()
+	labelSelector = labelSelector.Add(*labelMatch)
+	if err := clt.List(ctx, vsphereClusters, &client.ListOptions{LabelSelector: labelSelector, Namespace: cluster.Namespace}); err != nil {
+		return nil, err
+	}
+	if len(vsphereClusters.Items) != 1 {
+		return nil, fmt.Errorf("expected to find 1 VSphereCluster object for label key %s and value %s but found %d",
+			clusterapiv1beta1.ClusterLabelName, cluster.Name, len(vsphereClusters.Items))
+	}
+	return &vsphereClusters.Items[0], nil
+}
+
+// GetControlPlaneVsphereMachineTemplate gets the VsphereMachineTemplate CR of control plane
+func GetControlPlaneVsphereMachineTemplate(ctx context.Context, clt client.Client, cluster *clusterapiv1beta1.Cluster) (*capvv1beta1.VSphereMachineTemplate, error) {
+
+	vSphereMachineTemplates := &capvv1beta1.VSphereMachineTemplateList{}
+	labelMatch, err := labels.NewRequirement(clusterapiv1beta1.ClusterLabelName, selection.Equals, []string{cluster.Name})
+	if err != nil {
+		return nil, err
+	}
+	labelNotMatch, err := labels.NewRequirement(clusterapiv1beta1.ClusterTopologyMachineDeploymentLabelName, selection.DoesNotExist, nil)
+	if err != nil {
+		return nil, err
+	}
+	labelSelector := labels.NewSelector()
+	labelSelector = labelSelector.Add(*labelMatch, *labelNotMatch)
+	if err := clt.List(ctx, vSphereMachineTemplates, &client.ListOptions{LabelSelector: labelSelector, Namespace: cluster.Namespace}); err != nil {
+		return nil, err
+	}
+	if len(vSphereMachineTemplates.Items) != 1 {
+		return nil, fmt.Errorf("expected to find 1 vSphereMachineTemplate object with requirements %s in namespace %s but found %d",
+			labelSelector, cluster.Namespace, len(vSphereMachineTemplates.Items))
 	}
-	return vsphereCluster, nil
+	return &vSphereMachineTemplates.Items[0], nil
```

Comment:
How does this ensure you have the control plane machine template?

Reply:
You really need to do something like "KubeadmControlPlaneForCluster" , then introspect that and follow the infrastructureRef

Reply:
2 labels are used to filter out only control plane machine template:

- `cluster.x-k8s.io/cluster-name` = Cluster Name.
- `topology.cluster.x-k8s.io/deployment-name` does not exist.

Yes, infrastructureRef is more clear but increase UX complexity because users need to understand it and specify it correctly. Leave this to @lubronzhan because it 's about a design change.

Reply:
Thanks, got your point. I can get it from Cluster spec.controlPlaneRef -> KubeadmControlPlane spec.machineTemplate